[Weather Retrieval] Tour weather data improvements and new vendor addition (OpenWeatherMap) #479

FJBDev opened the pull request:
[Enhancement](https://github.com/mytourbook/mytourbook/issues/479)

wolfgang-ch commented:
It takes some time to review 113 files, until now I reviewed all collappsed folders 

![image](https://user-images.githubusercontent.com/1283445/156892958-386b30bb-9d85-4986-9df7-366bc50ea4c3.png)

FJBDev commented on bundles/net.tourbook/src/net/tourbook/database/TourDatabase.java:
```diff
@@ -8939,6 +8965,80 @@ private int updateDb_045_To_046(final Connection conn, final SplashManager splas
+
+         stmtUpdate = conn.prepareStatement(UI.EMPTY_STRING
+
+               + "UPDATE " + TABLE_TOUR_DATA //                      //$NON-NLS-1$
```
@wolfgang-ch 
This is the logic I set in place when the upgrade happens (46 ->47) to differentiate data from the device than data previously retrieved from the provider.
Thoughts ?

wolfgang-ch replied:
For me it looks OK

FJBDev replied:
In the case where a tour has weather data manually entered/from provider but also has a `temperatureSerie`, after the `46 -> 47` database migration, the device average, max & min temperature values will have the same values as in the average/max & min from provider.
But the user can always recalculate the device values with this existing contextual menu option:
![image](https://user-images.githubusercontent.com/25235665/157502251-762ee996-66d1-4287-b492-674f5f0e0b8e.png)

wolfgang-ch replied:
>In the case where a tour has weather data manually entered/from provider but also has a temperatureSerie, after the 46 -> 47 database migration, the device average, max & min temperature values will have the same values as in the average/max & min from provider.

This would be a bad behavior, after testing, the device avg/min/max are still the same as before.

The device avg/min/max values must be set only from the temperature serie values and in a few cases it is set from the device import

After manual edit

![image](https://user-images.githubusercontent.com/1283445/157813831-c7cbaf47-6062-41ca-b701-d7110d485238.png)

After loading weather data

![image](https://user-images.githubusercontent.com/1283445/157814640-b65d3578-3844-4475-9c4e-9a95a1f6ddd9.png)

FJBDev replied:
The only solution I see would be to trigger the computation of the device avg/min/max when performing the 46 -> 47 database upgrade but is that something possible during a database upgrade ??

FJBDev replied:
or maybe you meant they should be set to 0 and then recomputed manually by the user ?

wolfgang-ch replied:
I do not really understand what should be set and when?

Maybe the user can decide in preferences when to set/overwrite values?

>The only solution I see would be to trigger the computation of the device avg/min/max when performing the 46 -> 47 database upgrade but is that something possible during a database upgrade ??

This `net.tourbook.database.TourDatabase.updateDb_042_To_043_DataUpdate_Concurrent()` is modifying tour data during a db update but concurrently, otherwise it is very slow

FJBDev replied:
> I do not really understand what should be set and when?
> 

You said this yesterday so i understood that you wanted me to recompute the temperature values from the device after the update ? or did I misunderstand ??
>This would be a bad behavior, after testing, the device avg/min/max are still the same as before.


> This `net.tourbook.database.TourDatabase.updateDb_042_To_043_DataUpdate_Concurrent()` is modifying tour data during a db update but concurrently, otherwise it is very slow

I will look at that when you confirm what needs to be done.

wolfgang-ch replied:
I think there is some misunderstanding, this conversation started from your comment:

>In the case where a tour has weather data manually entered/from provider but also has a temperatureSerie, after the 46 -> 47 database migration, **the device average, max & min temperature values will have the same values as in the average/max & min from provider.**

Then I tested it but it was not as you had written (min/max/avg device data were still available), which was OK for me and there was no need to modify it.

FJBDev replied:
Maybe there is a misunderstanding, I'm going to try to explain better:

Here is the scenario of what will happen when going from DB 46 -> 47 with the current database code SQL query below):

1) If the tour has the boolean `isWeatherDataFromProvider` set to **true**, then the **current** (DB 46) values are copied into the new fields weather_Temperature_Average, Max & Min FROM the older weather_Temperature_Average_Device, Max, Min (Remember, in DB 46, there was only 1 temperature field that could have the value from the temperatureserie or manually entered from provider so we need some logic to know which value to put in which field)

2) If the tour has the boolean `isWeatherDataFromProvider` set to **false**, then we keep the values in weather_Temperature_Average_Device as we assume those values are from the device, so there is nothing to do

- If scenario 1) happens, then the values in weather_Temperature_Average & weather_Temperature_Average_Device are the same after the update but don't necessarily represent the actual value from the device. But like I mentioned [above](https://github.com/mytourbook/mytourbook/pull/483#discussion_r822930455), the user can recompute the device value manually or I can inspire myself from your function `updateDb_042_To_043_DataUpdate_Concurrent` to recompute the temperature values from the device with the TourData function `computeAvg_Temperature()`

Does it make sense ? If yes, what do you think ?


```
"UPDATE TABLE_TOUR_DATA 
SET                                  
 weather_Temperature_Average = weather_Temperature_Average_Device,
 weather_Temperature_Max = weather_Temperature_Max_Device,
 weather_Temperature_Min = weather_Temperature_Min_Device
 WHERE isWeatherDataFromProvider=true"
```

wolfgang-ch replied:
I overlooked the `WHERE isWeatherDataFromProvider=true"` and was also wondering, that my device values were not overwritten, sorry about that.

For scenario 1) you could either implement the concurrent tour data update solution (with the already existing showcase it should not too much work) or you should update the readme in the **Changes** section 

With your detailed description, now I could "fix" my thoughts :-)

FJBDev replied:
Ah ok, that makes sense now.

I have implemented the concurrent tour data update solution as I was able to inspire myself from `updateDb_042_To_043_DataUpdate_Concurrent`. Like you said, it was not too much work

FJBDev replied:
Fixed in my [PR](https://github.com/mytourbook/mytourbook/pull/485)

wolfgang-ch reviewed (changes requested):
This new weather provider is a useful addition (when I would know which weather station is used)

![image](https://user-images.githubusercontent.com/1283445/156918570-51638d57-fa7b-4b8f-a07a-24376b14cad6.png)

The wind speed/direction is wrong for 05/03 because is was from 45° during the tour and was also much higher than 3 km/h (blue line)

![image](https://user-images.githubusercontent.com/1283445/156918683-5c5bb5cc-97df-4753-b5f2-9da804877ca7.png)

FJBDev commented on bundles/net.tourbook/src/net/tourbook/ui/views/tourDataEditor/TourDataEditorView.java:
```diff
@@ -6228,6 +6367,7 @@ private void enableActions() {
                   && (_isManualTour == false));
       _actionToggleReadEditMode.setEnabled(isCellEditorInactive && isTourInDb);
 
+      final boolean canEdit = _isEditMode && isTourInDb;
```
I think Eclipse did that. Eclipse sometimes changes lines without notice. I spent 30+ mins reverting back `TourData.java`, `TourDataEditorView.java`, `TourDatabase.java` because a lot of lines were misplaced. I still don't understand how to prevent Eclipse from doing that and it is annoying. (I don't have this problem in any other Dev tools like Visual Studio etc...)

FJBDev replied:
Fixed

wolfgang-ch replied:
When fields are reformatted in Eclipse, only this helps:
* Add // comments for empty lines
* Format manually with the corresponding tags

When a method is commented out, then Eclipse is sometimes moving this comment to another method position, so I try to not comment it out and add a _UPPERCASE_POSTFIX_TAG to keep it

wolfgang-ch replied:
> I think Eclipse did that. Eclipse sometimes changes lines without notice. I wasted 30+ mins reverting back `TourData.java`, `TourDataEditorView.java`, `TourDatabase.java` because a lot of lines were misplaced. I still don't understand how to prevent Eclipse from doing that and it is annoying. (I don't have this problem in any other Dev tools like Visual Studio etc...)

As I already said, since a while I review all code before I do a commit, this takes some time but it prevents wrong formatting and it also helps to remove temporarily comments

FJBDev commented on bundles/net.tourbook/src/net/tourbook/ui/views/collateTours/TVICollatedTour.java:
```diff
@@ -53,9 +53,10 @@
-            + "AVG( CASE WHEN AvgTemperature = 0   THEN NULL ELSE DOUBLE(AvgTemperature) / TemperatureScale END )," + NL //   12 //$NON-NLS-1$
-            + "AVG( CASE WHEN WEATHERWINDDIR = 0   THEN NULL ELSE WEATHERWINDDIR END )," + NL //                              13 //$NON-NLS-1$
-            + "AVG( CASE WHEN WEATHERWINDSPD = 0   THEN NULL ELSE WEATHERWINDSPD END )," + NL //                              14 //$NON-NLS-1$
+            + "AVG( CASE WHEN weather_Temperature_Average_Device = 0   THEN NULL ELSE DOUBLE(weather_Temperature_Average_Device) / TemperatureScale END )," //$NON-NLS-1$
+            + NL //   12
```
I thought the same thing and didn't know what best solution was

FJBDev commented on bundles/net.tourbook/src/net/tourbook/ui/views/calendar/CalendarProfileManager.java:
```diff
@@ -3114,44 +3160,45 @@ private static void parse_200_Calendars(final XMLMemento xmlRoot,
 
             final String xmlProfileType = xmlProfile.getType();
 
-            if (xmlProfileType.equals(TAG_CALENDAR)) {
+            if (!xmlProfileType.equals(TAG_CALENDAR)) {
+               continue;
+            }
```
I will revert

FJBDev replied:
Fixed

FJBDev commented on bundles/net.tourbook/src/net/tourbook/ui/action/ActionAdjustTemperature.java:
```diff
@@ -65,7 +65,8 @@ public void run() {
       }
 
-      final DialogAdjustTemperature_Wizard wizard = new DialogAdjustTemperature_Wizard(selectedTours, _tourProvider);
+      final DialogAdjustTemperature_Wizard wizard =
+            new DialogAdjustTemperature_Wizard(selectedTours, _tourProvider);
```
I dont know if I did this or if Eclipse did. I will revert

FJBDev replied:
Fixed

FJBDev commented on bundles/net.tourbook/src/net/tourbook/data/TourData.java:
```diff
@@ -513,24 +514,28 @@
     * manually entered or from the device.
     */
-   private boolean               isWeatherDataFromApi = false;
+   private boolean               isWeatherDataFromProvider = false;
```
I inspired myself from other lines in [TourData.java](https://github.com/mytourbook/mytourbook/blob/e098eeb1ff2825581ef67d21ea3c337a5ff415a7/bundles/net.tourbook/src/net/tourbook/data/TourData.java#L464)

I will remove

FJBDev replied:
Fixed

wolfgang-ch replied:
This line https://github.com/mytourbook/mytourbook/blob/e098eeb1ff2825581ef67d21ea3c337a5ff415a7/bundles/net.tourbook/src/net/tourbook/data/TourData.java#L464 contains a comment for the 0 value, then I think it is OK to set a 0 value. 

In the beginning of Java programming, which was for me also the start of MT, I didn't know the default values, but after 15+ years, I know them :-)

FJBDev commented on bundles/net.tourbook/META-INF/MANIFEST.MF:
```diff
@@ -87,13 +89,16 @@ Import-Package: com.fasterxml.jackson.core.type,
  net.tourbook.web.preferences,
  org.apache.commons.collections4,
  org.apache.commons.lang3.text,
+ org.apache.http.client.utils;version="4.5.13",
```
I think Eclipse added automatically, I can try to remove it.

FJBDev replied:
Removed

FJBDev commented on bundles/net.tourbook/src/net/tourbook/ui/views/rawData/DialogReimportTours.java:
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (C) 2020, 2021 Frédéric Bard
+ * Copyright (C) 2020, 2022 Frédéric Bard
  *
```
I can't reproduce the issue. Can you tell me the scenario to reproduce it ?

FJBDev replied:
Fixed

FJBDev commented on bundles/net.tourbook/src/net/tourbook/tour/TourManager.java:
```diff
@@ -2715,49 +2738,57 @@ private static void removeTourPauses(final TourData tourData,
+               });
+      }
+
+      TourLogManager.subLog_INFO(String.format(
+            LOG_RETRIEVE_WEATHER_DATA_002_END,
+            (System.currentTimeMillis() - start) / 1000.0));
+
```
Actually, with OpenWeatherMap, the result gives the hourly data so I will improve it by taking al the hours from the tour start to the tour end times.

FJBDev replied:
>Could it be possible to show also the weather station in the log?

OpenWeatherMap doesn't seem to give back the weather station in their API response. You can look at an example [here](https://github.com/mytourbook/mytourbook/blob/fb-improvements/bundles/net.tourbook.tests/src/test/data/weather/openweathermap/files/LongsPeak-Manual-OpenWeatherMapResponse.json) but I don't see such data.

FJBDev replied:
OpenWeatherMap weather retrieval is onw improved to only take the values included between the tour start and end times

wolfgang-ch replied:
> Actually, with OpenWeatherMap, the result gives the hourly data so I will improve it by averaging all the data between the tour start to the tour end times.

Can you log all the hour/temperature values which are used to compute the average values, that someone know how it was calculated, otherwise this is a blackbox

FJBDev replied:
>>Can you log all the hour/temperature values which are used to compute the average values, that someone know how it was calculated, otherwise this is a blackbox

I can. Do you have a favorite format to log those values, or anything will do ?

FJBDev replied:
I added an option for the user to display the "full log" when retrieving weather data:

- The new preference option that is displayed for all the providers
![image](https://user-images.githubusercontent.com/25235665/157127058-4090929f-da2b-4cf0-8b62-bc5a83f9e809.png)

- An example of "full log" when using OpenWeatherMap
![image](https://user-images.githubusercontent.com/25235665/157127101-6f490902-1e7f-4dc9-b4c5-5f976ae57c4e.png)

FJBDev replied:
> > Could it be possible to show also the weather station in the log?
> 
> OpenWeatherMap doesn't seem to give back the weather station in their API response. You can look at an example [here](https://github.com/mytourbook/mytourbook/blob/fb-improvements/bundles/net.tourbook.tests/src/test/data/weather/openweathermap/files/LongsPeak-Manual-OpenWeatherMapResponse.json) but I don't see such data. However, it seems that [WWO gives the location information](https://github.com/mytourbook/mytourbook/blob/8bca5dee7aa5855afec3be2e959233d1329dab14/bundles/net.tourbook.tests/src/test/data/weather/worldweatheronline/files/LongsPeak-Manual-WorldWeatherOnlineResponse.json#L26) so I can add that for WorldWeatherOnline

I added the weather station information (for WorldWeatherOnline since OpenWeatherMap doesn't have such information) in the log when the "full log" is checked. The weather station information is a clickable link that opens the weather station website as the example below:

![image](https://user-images.githubusercontent.com/25235665/157166196-4531f13e-16d3-4c78-b8b1-81f646e8e05f.png)

FJBDev commented on bundles/net.tourbook/src/net/tourbook/tour/TourInfoUI.java:
```diff
@@ -869,6 +871,15 @@ private void createUI_44_Weather(final Composite parent) {
 
       createUI_Label(parent, UI.UNIT_LABEL_TEMPERATURE);
 
+      /*
+       * Temperature from device
+       */
+      createUI_Label(parent, Messages.Tour_Tooltip_Label_Temperature_FromDevice);
+
+      _lblTemperature_FromDevice = createUI_LabelValue(parent, SWT.TRAIL);
+
+      createUI_Label(parent, UI.UNIT_LABEL_TEMPERATURE);
+
```
I will improve the UI based on your suggestions

FJBDev replied:
>When possible, then the manual/provider temperature should only be visible when these values are set, maybe also for the device temperature but this may needs more changes

That's hard because by default, the temperature value is 0 which is a possible value in real life. So how can we know that those values are "set" ??

FJBDev replied:
Tooltip improved. Also, now the average temperature from the device is only displayed if the `temperatureSerie` is not empty

![image](https://user-images.githubusercontent.com/25235665/156966044-3026c7f8-23cc-4273-aa79-a0164c5242ea.png)

wolfgang-ch replied:
Almost perfect, the weather values are looking a bit chaotic (similar like the weather sometimes is), I will try to improve it

FJBDev commented on bundles/net.tourbook/src/net/tourbook/ui/views/tourDataEditor/TourDataEditorView.java:
```diff
@@ -6385,28 +6525,23 @@ private void enableControls() {
-      _spinWeather_Temperature_WindChill.setEnabled(canEdit);
-      _spinWeather_Wind_DirectionValue.setEnabled(canEdit);
+      _spinWeather_Wind_DirectionValue.setEnabled(canEdit &&
+            _comboWeather_WindDirectionText.getSelectionIndex() > 0);
```
Will do

FJBDev replied:
Fixed

FJBDev commented on bundles/net.tourbook/src/net/tourbook/ui/views/rawData/DialogDeleteTourValues.java:
```diff
@@ -723,7 +726,14 @@ protected void okPressed() {
          DialogUtils.addTourValueTypeFromCheckbox(_chkData_RunningDynamics, TourValueType.TIME_SLICES__RUNNING_DYNAMICS, tourValueTypes);
          DialogUtils.addTourValueTypeFromCheckbox(_chkData_Swimming, TourValueType.TIME_SLICES__SWIMMING, tourValueTypes);
-         DialogUtils.addTourValueTypeFromCheckbox(_chkData_Temperature, TourValueType.TIME_SLICES__TEMPERATURE, tourValueTypes);
+         DialogUtils.addTourValueTypeFromCheckbox(
+               _chkData_Temperature,
+               TourValueType.TIME_SLICES__TEMPERATURE,
+               tourValueTypes);
+         DialogUtils.addTourValueTypeFromCheckbox(
+               _chkData_Temperature_FromDevice,
+               TourValueType.TIME_SLICES__TEMPERATURE_FROMDEVICE,
+               tourValueTypes);
```
Maybe you have a wide screen, I need to scroll back and forth on my screen. But I can revert.

FJBDev replied:
Fixed

wolfgang-ch replied:
The line formatting depends also on the surrounding code, when there is similar codes around then I try to format it like the surrounding code. 
For fields this is done automatically but not for other code lines

FJBDev commented on bundles/net.tourbook/src/net/tourbook/ui/views/tagging/TaggingView.java:
```diff
@@ -145,20 +143,21 @@
    static final int                  TAG_VIEW_LAYOUT_FLAT                   = 0;
-   static final int                  TAG_VIEW_LAYOUT_HIERARCHICAL           = 10;
 
+   static final int                  TAG_VIEW_LAYOUT_HIERARCHICAL           = 10;
    private static final NumberFormat _nf0                                   = NumberFormat.getNumberInstance();
+
    private static final NumberFormat _nf1                                   = NumberFormat.getNumberInstance();
```
Another one probably made by Eclipse automatically. I will revert manually.

FJBDev replied:
Fixed

FJBDev commented on bundles/net.tourbook/src/net/tourbook/ui/views/tourDataEditor/TourDataEditorView.java:
```diff
@@ -588,7 +590,7 @@
    private ColumnManager            _swimSlice_ColumnManager;
    private SwimSlice_TourViewer     _swimSlice_TourViewer         = new SwimSlice_TourViewer();
 
-   private FormToolkit              _tk;
+   private FormToolkit              _formToolkit;
```
I changed them because `tk` means nothing to me. Does it mean "Telekommunikation" ?  "[to come](https://en.wikipedia.org/wiki/To_come_(publishing))" ??? `_pc` does it mean computer like a PC ?? I know you've been using them for 10 years so probably it means "something" for you but not for me. But since you ask, I will revert.

FJBDev replied:
Reverted to `_tk`

wolfgang-ch replied:
_tk ... toolkit
_pc ... pixel converter

FJBDev commented:
> This new weather provider is a useful addition (when I would know which weather station is used)
> 
See my answer [here](https://github.com/mytourbook/mytourbook/pull/483#discussion_r820267219)
> 
> The wind speed/direction is wrong for 05/03 because is was from 45° during the tour and was also much higher than 3 km/h (blue line)
> 

It's possible that the data returned is different that other weather apps. If you send me the tour, I can debug to see if it's a bug in my code or if the data returned is the problem.
Did you try to compare with WorldWeatherOnline (you can get a free trial API just for testing) ?

FJBDev commented on bundles/net.tourbook.export/format-templates/gpx-1.0.vm:
```diff
@@ -47,11 +47,11 @@
          <mt:dpTolerance>$tourData.DpTolerance</mt:dpTolerance>
 ## weather
-         <mt:temperature>$tourData.AvgTemperature</mt:temperature>
+         <mt:temperature>$tourData.weather_Temperature_Average_Device</mt:temperature>
```
@wolfgang-ch 
Thanks to the unit tests 😄  that failed, I was able to update this file, otherwise I would have broken this export GPX feature.

wolfgang-ch reviewed (changes requested):
A few more issues.

The log is now helpful when the formatting is tweacked :-)

I would use String.format(...) to force the columns

![image](https://user-images.githubusercontent.com/1283445/157203831-e58a4157-dfed-4756-8c6e-4f434c196c02.png)

FJBDev commented on bundles/net.tourbook/src/net/tourbook/importdata/DialogEasyImportConfig.java:
```diff
@@ -3393,20 +3340,12 @@ private void enable_IL_Controls() {
-                        label.setImage(null);
-                     }
+                     label.setImage(TourTypeImage.getTourTypeImage(tourTypeId));
                   }
```
Because if you look closely at the current file, you test `isILSelected` twice but it doesn't seem like it has changed in the meantime. That seems redundant and not efficient to test a boolean twice no ?

![image](https://user-images.githubusercontent.com/25235665/157262115-c7710eee-8804-4c66-a76e-a01623b32d31.png)

FJBDev replied:
For me it was easier because it's [SonarLint ](https://www.sonarlint.org/eclipse)that told me about it 😄 

There's also this issue that it underlined:

![image](https://user-images.githubusercontent.com/25235665/157344242-2b683b44-617c-47b8-80f7-18129600aa9c.png)

wolfgang-ch replied:
Fixed in https://github.com/mytourbook/mytourbook/commit/f15aafea8c5008c20d948c9f04f34be79051b263

FJBDev commented on bundles/net.tourbook/src/net/tourbook/importdata/DialogEasyImportConfig.java:
```diff
@@ -4810,7 +4717,7 @@ private void update_UI_From_Model_IL() {
                   spinnerAvgSpeed.setData(DATA_KEY_SPEED_TOUR_TYPE_INDEX, speedTTIndex);
                   comboCadence.setData(DATA_KEY_SPEED_TOUR_TYPE_INDEX, speedTTIndex);
-                  _actionTTSpeed_Delete[speedTTIndex].setData(DATA_KEY_SPEED_TOUR_TYPE_INDEX, speedTTIndex);
+                  _actionTTSpeed_Delete[speedTTIndex].setData(speedTTIndex);
```
Because the [setData ](https://github.com/mytourbook/mytourbook/blob/4a4bf14e8fbe9c72b553117abf3427354f7859eb/bundles/net.tourbook/src/net/tourbook/importdata/DialogEasyImportConfig.java#L424)function doesn't use this parameter `key`

wolfgang-ch replied:
I didn't saw that this is a private function, because calling `setData()` for a widget could be dangerous as it could overwrite data, therefore I normally use a key

FJBDev replied:
so you want me to revert it ?

FJBDev commented on bundles/net.tourbook/src/net/tourbook/tour/DialogQuickEdit.java:
```diff
@@ -188,7 +189,7 @@ private Composite createSection(final Composite parent, final String title, fina
       final Section section = _tk.createSection(parent, //
       //Section.TWISTIE |
 //            Section.SHORT_TITLE_BAR
-            Section.TITLE_BAR
+            ExpandableComposite.TITLE_BAR
```
Maybe another Eclipse optimization ? I can revert if you want

FJBDev replied:
Fixed

FJBDev commented on bundles/net.tourbook/src/net/tourbook/Messages.java:
```diff
@@ -930,6 +930,7 @@
    public static String        External_Link_MyTourbook;
    public static String        External_Link_MyTourbook_TourChartSmoothing;
    public static String        External_Link_Weather_ApiSignup;
+   public static String        External_Link_Weather_OpenWeatherMap;
```
I will fix. Maybe you should fix all the previous keys so we don't get inspired from old bad code.

FJBDev replied:
Fixed for all the occurrences I have found: WorldWeatherOnline, OpenWeatherMap, Strava, Suunto.....
Let me know if you find more

wolfgang-ch replied:
Removed "my" external links in https://github.com/mytourbook/mytourbook/commit/8078c7336e7303d3ef956e9a673f5a6fb9a00bb7

FJBDev commented on bundles/net.tourbook/src/net/tourbook/importdata/RawDataManager.java:
```diff
@@ -1450,9 +1450,18 @@ private TourData createTourDataDummyClone(final List<TourValueType> tourValueTyp
 
-            if (isEntireTour_OR_AllTimeSlices || tourValueType == TourValueType.TIME_SLICES__TEMPERATURE) {
+            if (isEntireTour_OR_AllTimeSlices ||
+                  tourValueType == TourValueType.TIME_SLICES__TEMPERATURE) {
+
+               tourDataDummyClone.setWeather_Temperature_Average(
+                     oldTourData.getWeather_Temperature_Average());
+            }
+
+            if (isEntireTour_OR_AllTimeSlices ||
+                  tourValueType == TourValueType.TIME_SLICES__TEMPERATURE_FROMDEVICE) {
 
-               tourDataDummyClone.setAvgTemperature(oldTourData.getAvgTemperature());
+               tourDataDummyClone.setWeather_Temperature_Average_Device(
+                     oldTourData.getWeather_Temperature_Average_Device());
```
Fixed

FJBDev commented on bundles/net.tourbook/src/net/tourbook/ui/Messages.java:
```diff
@@ -599,12 +599,15 @@
    public static String        Tour_Tooltip_Label_Weather;
    public static String        Tour_Tooltip_Label_WindDirection;
    public static String        Tour_Tooltip_Label_WindSpeed;
+   public static String        Tour_Tooltip_Value_AverageTemperatures;
+   public static String        Tour_Tooltip_Value_AverageTemperatures_Tooltip;
```
Fixed

FJBDev commented on bundles/net.tourbook/src/net/tourbook/weather/openweathermap/OpenWeatherMapRetriever.java:
```diff
@@ -0,0 +1,186 @@
+      if (_timeMachineResult == null) {
+         return false;
+      }
+
+      tour.setIsWeatherDataFromProvider(true);
+      tour.setWeather(_timeMachineResult.getWeatherDescription());
+      tour.setWeather_Clouds(_timeMachineResult.getWeatherType());
+      tour.setWeather_Temperature_Average(_timeMachineResult.getTemperatureAverage(tour));
+      tour.setWeather_Wind_Speed(_timeMachineResult.getWindSpeed(tour));
+      tour.setWeather_Wind_Direction(_timeMachineResult.getWindDirection(tour));
+      tour.setWeather_Humidity((short) _timeMachineResult.getAverageHumidity(tour));
+      tour.setWeather_Precipitation(_timeMachineResult.getPrecipitation(tour));
+      tour.setWeather_Pressure((short) _timeMachineResult.getAveragePressure(tour));
+      tour.setWeather_Temperature_Max(_timeMachineResult.getTemperatureMax(tour));
+      tour.setWeather_Temperature_Min(_timeMachineResult.getTemperatureMin(tour));
+      tour.setWeather_Temperature_WindChill(_timeMachineResult.getWindChill(tour));
+
```
I think you meant to start with SET_FORMATTING_**OFF** right ?? In your picture you start with **ON**

Fixed
![image](https://user-images.githubusercontent.com/25235665/157313832-2ea53da0-1fa1-4f30-996d-5e7ada0b0ff7.png)

FJBDev commented on bundles/net.tourbook/src/net/tourbook/Messages.java:
```diff
@@ -2970,6 +2990,8 @@
 
    public static String        ui_tour_not_defined;
 
+   public static String        Weather_Provider_None;
```
Fixed

FJBDev commented on bundles/net.tourbook/src/net/tourbook/tour/DialogQuickEdit.java:
```diff
@@ -988,12 +959,20 @@ private void updateUIFromModel() {
+            final int weatherWindDirectionDegree = weatherWindDirection * 10;
+            _spinWeather_Wind_DirectionValue.setSelection(weatherWindDirectionDegree);
+            _comboWeather_Wind_DirectionText.select(UI.getCardinalDirectionTextIndex((int) (weatherWindDirectionDegree / 10.0f)));
+            _spinWeather_Wind_DirectionValue.setEnabled(true);
```
Fixed

FJBDev commented:
> A few more issues.
> 
> The log is now helpful when the formatting is tweacked :-)
> 
> I would use String.format(...) to force the columns
> 
> ![image](https://user-images.githubusercontent.com/1283445/157203831-e58a4157-dfed-4756-8c6e-4f434c196c02.png)

I improved it but I am reluctant to use `String.format` because the length of the columns could be very different depending the language used. Is there an example in MT where this is used and the column look the same for ANY languages used ?

![image](https://user-images.githubusercontent.com/25235665/157321485-2eefdb06-3ed5-463e-a893-4157900011b4.png)

FJBDev commented on bundles/net.tourbook/src/net/tourbook/ui/views/tourDataEditor/TourDataEditorView.java:
```diff
@@ -8845,12 +8933,20 @@ private void updateUI_Tab_1_Tour() {
+         final int weatherWindDirectionDegree = weatherWindDirection * 10;
+         _spinWeather_Wind_DirectionValue.setSelection(weatherWindDirectionDegree);
+         _comboWeather_WindDirectionText.select(UI.getCardinalDirectionTextIndex((int) (weatherWindDirectionDegree / 10.0f)));
+         _spinWeather_Wind_DirectionValue.setEnabled(true);
```
Fixed.

wolfgang-ch commented:
>[](https://user-images.githubusercontent.com/25235665/157313832-2ea53da0-1fa1-4f30-996d-5e7ada0b0ff7.png)I think you meant to start with SET_FORMATTING_OFF right ?? In your picture you start with ON

Yes, I moved the wrong line for this demo, my template inserts it correctly

![image](https://user-images.githubusercontent.com/1283445/157380877-264c103a-e141-47f9-b81d-ac96202cdb5e.png)

wolfgang-ch commented on bundles/net.tourbook/src/net/tourbook/ui/views/tourDataEditor/TourDataEditorView.java:
```diff
@@ -9032,6 +9182,7 @@ private void updateUI_Tab_1_Tour() {
        */
       _tourContainer.layout(true);
 
+      enableControls();
```
`enableControls()` is already called after this method, when there is a need to call `enableControls()`, then it should be called outside of `updateUI_Tab_1_Tour()`

![image](https://user-images.githubusercontent.com/1283445/157382283-5ff378de-1d36-4539-85d4-0aa7f2b602f3.png)

FJBDev replied:
Indeed, I didn't see that. Removed.

wolfgang-ch commented on bundles/net.tourbook/src/net/tourbook/weather/WeatherUtils.java:
```diff
@@ -0,0 +1,326 @@
+    *           in mm
+    * @param time
+    *           in epoch seconds
+    * @param timeZonedId
+    *           Example: 17h(15.0°C, feels like 15.0°C, 5.0km/h from 68°, humidity 37%,
+    *           precipitation 0.0mm)
```
Wrong location for this comment

![image](https://user-images.githubusercontent.com/1283445/157383418-131714db-3f11-4dad-b24d-d1d14892273f.png)

FJBDev replied:
Fixed

wolfgang-ch commented:
I cannot comment on this comment https://github.com/mytourbook/mytourbook/pull/483#discussion_r822028590, so here is my answer

> so you want me to revert it ?

No

wolfgang-ch commented:
> > A few more issues.
> > The log is now helpful when the formatting is tweacked :-)
> > I would use String.format(...) to force the columns
> > ![image](https://user-images.githubusercontent.com/1283445/157203831-e58a4157-dfed-4756-8c6e-4f434c196c02.png)
> 
> I improved it but I am reluctant to use `String.format` because the length of the columns could be very different depending the language used (and therefore maybe it could look good in english but maybe not as good in another language). Is there an example in MT where this is used and the column look the same for ANY languages used ?
> 
> ![image](https://user-images.githubusercontent.com/25235665/157321485-2eefdb06-3ed5-463e-a893-4157900011b4.png)

This looks already good now. 

I thought for text you format it with `%s` for values you can set a fixed number of digits. When you display the log with the browser, it is not using a mono space font but monospace is used when the log is displayed as normal text

wolfgang-ch commented:
> Is there an example in MT where this is used and the column look the same for ANY languages used ?

This example is using different text lengths

`net.tourbook.database.DialogCompressDatabase.createLog_BeforeCompressed(...)`

wolfgang-ch reviewed (commented):
Almost done :-)

FJBDev commented:
> Yes, I moved the wrong line for this demo, my template inserts it correctly
> 
> ![image](https://user-images.githubusercontent.com/1283445/157380877-264c103a-e141-47f9-b81d-ac96202cdb5e.png)

I need to read the doc and see if it explains how to install those useful templates that I don't use yet.

FJBDev commented on bundles/net.tourbook/src/net/tourbook/tour/DialogQuickEdit.java:
```diff
@@ -1007,17 +984,13 @@ private void updateUIFromModel() {
 
-         _spinWeather_Temperature_Avg.setDigits(1);
-         _spinWeather_Temperature_Avg.setSelection(Math.round(avgTemperature * 10));
+         enableControls();
```
I missed that one too. Fixed

FJBDev commented on bundles/net.tourbook/src/net/tourbook/tour/DialogQuickEdit.java:
```diff
@@ -798,7 +797,8 @@ private void displayCloudIcon() {
 
-      _spinWeather_Temperature_Avg.setEnabled(_tourData.temperatureSerie == null);
+      _spinWeather_Wind_DirectionValue.setEnabled(
+            _comboWeather_Wind_DirectionText.getSelectionIndex() > 0);
```
Another Eclipse automatic action. Fixed

FJBDev commented:
> Almost done :-)

I think I addressed all your code review/suggestions. Then I remembered that I needed to add the "snowfall" value in the weather retrieval. I have implemented that and updated the screenshots in the [Github enhancement.](https://github.com/mytourbook/mytourbook/issues/479)

wolfgang-ch reviewed (commented):
2 UI improvements

FJBDev commented on bundles/net.tourbook/src/net/tourbook/tour/TourInfoUI.java:
```diff
@@ -1736,20 +1737,28 @@ private void updateUI() {
+      // Average temperature
+      final float averageTemperature = UI.convertTemperatureFromMetric(
+            _tourData.getWeather_Temperature_Average());
+      final float temperatureFromDeviceValue = UI.convertTemperatureFromMetric(
+            _tourData.getWeather_Temperature_Average_Device());
+
+      final String temperatureFromDevice = _tourData.temperatureSerie != null && _tourData.temperatureSerie.length > 0
+            ? _nf1.format(temperatureFromDeviceValue) + UI.UNIT_LABEL_TEMPERATURE
+            : UI.EMPTY_STRING;
```
Sure. I fixed it and added 2 spaces between the "-" and the device value

![image](https://user-images.githubusercontent.com/25235665/157698430-38d9dc88-9411-431f-9fdc-18cba8b2a3d0.png)

FJBDev commented on bundles/net.tourbook/src/net/tourbook/weather/WeatherUtils.java:
```diff
@@ -0,0 +1,342 @@
+            precipitation + UI.UNIT_LABEL_DISTANCE_MM_OR_INCH + UI.COMMA_SPACE +
+            snowFall + UI.UNIT_LABEL_DISTANCE_MM_OR_INCH +
+            UI.SYMBOL_BRACKET_RIGHT;
+
```
I added the fixed value digits but there is an issue with the log indent not working.
The code [here ](https://github.com/mytourbook/mytourbook/blob/937bbcf10e55d8b15888eb2d4569f49de1a32678/bundles/net.tourbook/src/net/tourbook/tour/TourLogManager.java#L217)says `Indent log message and show info icon` but it seems that the indentation works only for the first line . Am I missing something ??

![image](https://user-images.githubusercontent.com/25235665/157724066-c1f7487a-e178-43ef-960b-e665bceb3b71.png)

wolfgang-ch replied:
Improved logging in https://github.com/mytourbook/mytourbook/commit/3edb48e278c010d550dde160329673ab59846ff7

![image](https://user-images.githubusercontent.com/1283445/157846778-74e49047-e629-4701-8965-286e4b2e256a.png)

FJBDev replied:
That doesn't seem to work, it's not straight when I tested it
![image](https://user-images.githubusercontent.com/25235665/157893858-65af51fe-a5a8-40b9-bf1d-30257049d682.png)

wolfgang-ch replied:
You have not really formatted the values with fixed widths, fixed width formatting would be e.g. `%5.2f` but you formatted them with `%5s,  value` and the `value` is formatted with `value.toString()` which is not setting a fixed width

FJBDev replied:
Fixed in my [PR](https://github.com/mytourbook/mytourbook/pull/485)

![image](https://user-images.githubusercontent.com/25235665/158252940-4adf797a-6837-4ba5-9c4a-abeb7e458f19.png)

FJBDev commented on bundles/net.tourbook.common/src/net/tourbook/common/Messages.java:
```diff
@@ -276,15 +276,15 @@
+   public static String        Weather_Clouds_ScatteredShowers;
+   public static String        Weather_Clouds_SevereWeatherAlert;
+   public static String        Weather_Clouds_Snow;
+   public static String        Weather_Clouds_Sunny;
```
@wolfgang-ch 
Today i found and fixed those typo issues

wolfgang-ch replied:
A typo since 2010 fee487aba0a19dab8cbf29901bee8b114a896086

FJBDev replied:
Wow, impressive the nobody as found the typo before me

wolfgang-ch commented on bundles/net.tourbook/src/net/tourbook/tour/DialogQuickEdit.java:
```diff
@@ -81,7 +82,6 @@
-   private float                    _unitValueTemperature;
 
    /*
     * UI controls
```
Can you display the device avg temperature also in the quick edit dialog, when there is only one value, it is irritating

![image](https://user-images.githubusercontent.com/1283445/157815667-be94bf16-c4db-499c-999f-818c0083c10f.png)

FJBDev replied:
Sure, I will add that in my next PR

FJBDev replied:
Fixed in my [PR](https://github.com/mytourbook/mytourbook/pull/485)

![image](https://user-images.githubusercontent.com/25235665/158253032-35568e2d-abe0-48eb-8d3f-25e6f777e95b.png)
